### DISCUSSION TITLE
Jinja alert type error in percentage match rule

### DISCUSSION THREAD

**Discussion by firdausa-iprogrammer** [jertel/elastalert2#328] [Q&A] [answered]
Hi Team,

Getting the following error when I try to run a percentage match rule -
```
ERROR:elastalert:Uncaught exception running rule test_alerts/L1/generateOTP_v3-DXL202422-pct: unsupported operand type(s) for |: 'dict' and 'dict'
```

I need to do the following for the 'percentage' output from the rule -
Format it to only show 2 decimals(percentage_format_string already enabled for '%.2f') but it still has multiple decimals.

Error rule config
```
alert_text_type: alert_text_jinja
alert_subject: "L1/HR:DXL202422 generateOTP threshold: 5% current: {{percentage}}"
#alert_subject_args: ["percentage"]

alert_text: |
    Hi Team,
    For Vi Application DXL Error code DXL202422 generateOTP API Failure threshold value 5% breached to {{percentage}} in last one hour.
    SOR system share the Current Issue Analysis.
    Regards,
    Vi DevOps

#alert_text_args: ["value"]
```

Please advise.

**Comment by jertel** [comment, marked as answer]
Are there two questions being asked in this post?

**Reply by firdausa-iprogrammer** [reply]
@jertel -
Yes sorry -
The primary question is how to get only 2 decimal values in the "percentage" field.
The secondary question is about the jinja error, which was an attempt to solve the primary question myself.

So basically, if I can get only 2 decimals working using "alert_text_only" combined with "percentage_format_string" and "alert_text_args and alert_subject_args" that solves both questions.

However if I do need to switch to Jinja, then first I need to know why I'm getting the error and secondly how to get the "percentage" value to only 2 decimal places

**Reply by jertel** [reply]
The Jinja error would make sense if you were using an outdated version of Python. ElastAlert 2 requires Python 3.9 (See https://elastalert2.readthedocs.io/en/latest/running_elastalert.html#as-a-python-package).

**Reply by firdausa-iprogrammer** [reply]
Ah! That explains it I'm still on python 3.6.
Does this mean I will have to use "alert_text_jinja"?
Or is there some way to get only 2 decimal points in "alert_text_only"?

**Comment by firdausa-iprogrammer** [comment]
Hi @jertel,

You were spot on with the Jinja error, updated Python to 3.9.6 and now it's working.
This is what I'm using as the Jinja template -
```{{'{:.2f}'.format(percentage)}}```

It's working perfectly with only 2 decimal places.

If I could suggest an enhancement however, would it be possible to pass the percentage_format_string to the percentage field as well?
Currently it only applies to the match statement as far as I can tell.

```
def get_match_str(self, match):
        percentage_format_string = self.rules.get('percentage_format_string', None)
        message = 'Percentage violation, value: %s (min: %s max : %s) of %s items\n\n' % (
            percentage_format_string % (match['percentage']) if percentage_format_string else match['percentage'],
            self.rules.get('min_percentage'),
            self.rules.get('max_percentage'),
            match['denominator']
        )
	return message
```

**Comment by firdausa-iprogrammer** [comment]
I would also like to know how to format the field in the command alerter.

**Reply by firdausa-iprogrammer** [reply]
I think I figured it out -
```{percentage:.2f}```

Thanks a ton for everything else @jertel !!!
Please do consider my Enhancement request in the last reply

**Reply by nsano-rururu** [reply]
Is the following description helpful?

Update documentation for percentage_format_string and alerta_timeout
https://github.com/Yelp/elastalert/pull/2537/files
>.. note:: If you want to change the percentage format inside your alert, you must specify the format there cause ``percentage_format_string`` wont't be applied.
    Exemple: ``alerta_value: "%(percentage).2f%%"``

**Reply by firdausa-iprogrammer** [reply]
@nsano-rururu Definitely would be useful, also please add the description for the new-format style which I've used or if there's something better then please use that.
